kin-sdk/kin-sdk-lib/src/main/java/kin/sdk/transactiondata/Transaction.java

```diff
@@ -0,0 +1,189 @@
+import kin.sdk.internal.data.TransactionIdImpl;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.List;
+
+public abstract class Transaction {
+
+    private final kin.base.Transaction baseTransaction;
+
+    public Transaction(kin.base.Transaction baseTransaction) {
+        this.baseTransaction = baseTransaction;
+    }
+
+    /**
+     * Creates a <code>RawTransaction</code> or a <code>PaymentTransaction</code> instance from previously build
+     * <code>TransactionEnvelope</code> string representation.
+     *
+     * <p>Once a transaction is ready to be signed, the transaction object is wrapped in an object called
+     * a Transaction Envelope, which contains the transaction as well as a set of signatures.
+     * A Transaction Envelope can be represented as a string base-64 encoded.</p>
+     *
+     * @param transactionEnvelope a Base-64 encoded <code>TransactionEnvelope</code>
+     * @return a Transaction instance from a previously build transaction envelope.
+     */
+    public static Transaction decodeTransaction(String transactionEnvelope) throws DecodeTransactionException {
+        try {
+            kin.base.Transaction transaction = kin.base.Transaction.fromEnvelopeXdr(transactionEnvelope);
+            Operation[] operations = transaction.getOperations();
+            int paymentOperationCount = 0;
+            PaymentOperation paymentOperation = null;
+            for (Operation operation : operations) {
+                if (operation instanceof PaymentOperation) {
+                    paymentOperationCount++;
+                    paymentOperation = (PaymentOperation) operation;
+                }
+            }
+            if (paymentOperationCount == 1) {
+                return new PaymentTransaction(transaction, paymentOperation.getDestination().getAccountId(),
+                        new BigDecimal(paymentOperation.getAmount()), ((MemoText) transaction.getMemo()).getText());
+            } else {
+                return new RawTransaction(transaction);
+            }
+        } catch (IOException e) {
+            throw new DecodeTransactionException(e.getMessage(), e.getCause());
+        }
+    }
+
+    /**
+     * Creates a <code>RawTransaction</code> instance from previously build <code>TransactionEnvelope</code>
+     * <p> See {@link Transaction#decodeTransaction(String)} for more details</p>
+     *
+     * @param transactionEnvelope a Base-64 encoded <code>TransactionEnvelope</code>
+     */
+    public static RawTransaction decodeRawTransaction(String transactionEnvelope) throws DecodeTransactionException {
+        try {
+            kin.base.Transaction transaction = kin.base.Transaction.fromEnvelopeXdr(transactionEnvelope);
+            return new RawTransaction(transaction);
+        } catch (IOException e) {
+            throw new DecodeTransactionException(e.getMessage(), e.getCause());
+        }
+    }
+
+    public kin.base.Transaction baseTransaction() {
```

Comment:
hide by introducing a middle level class that will be used only internally (TransactionBase)